source/language/openpulse.rst

```diff
@@ -0,0 +1,742 @@
+        tx(d1, sig0, 3);
+
+.. _Timing:
+
+Timing
+------
+
+Each ``channel`` maintains its own "clock". When a signal is transmitted/received, the channel clock
+is played the clock for advances by the length of the signal (``signal.duration``).
+
+``delay`` and ``barrier`` instructions in OpenPulse resolve timing as in the qubit case outlined in the
+`Delays <delays.html>`_ section of the OpenQASM specification. At the pulse level, however,
+``delay`` and ``barrier`` take ``channels`` as input, rather than qubits.
+
+.. code-block:: c
+
+    signal sig = ...; // some 100dt pulse
+
+    defcal simultaneous_pulsed_gate $0 {
+        txchannel d0 = txch($0, "drive");
+        txchannel d1 = txch($1, "drive");
+        // sig begins playing on ``d0``
+        transmit(d0, sig, sig.duration);
+        // Delay d1 by 20dt
+        delay(d1, 20dt);
+        // Next signal on ``d1`` will start at ``20dt``. We now play ``sig`` on ``d1``. It trails the
+        // ``sig`` on ``d0`` by ``20dt``. We only play it on ``d1`` for ``80dt`` so it finishes at the
+        // same time as ``d0``. Note that the full signal will play on ``d0``, but only the first 80
+        // samples will play on ``d1``.
+        transmit(d1, sig, sig.duration-20dt);
+    }
+
+``defcal`` blocks have an implicit barrier on every channel used within the block,
+meaning that clocks are guaranteed to be aligned at the start of the block. Channel clock time persists
+between ``defcal``'s within a common program, so the first use of channels within a ``defcal``
+need not be at ``t=0``.
+
+.. code-block:: c
+
+    defcal cal1 $0, $1 {
+        txchannel d0 = txch($0, "drive");
+        txchannel d1 = txch($1, "drive");
+        signal sig1 = ...;  // some 100dt pulse
+        signal sig2 = shift(sig1, 20dt);  // shift to start at ``20dt``, length is only ``80dt``
+        transmit(d0, sig1, sig1.duration);
+        transmit(d1, sig2, sig2.duration);
+    }
+
+    defcal cal2 $0, $1 {
+        txchannel d0 = txch($0, "drive");
+        txchannel d1 = txch($1, "drive");
+        signal sig3 = ...; // some 50dt pulse
+        signal sig4 = ...; // some 75dt pulse
+        transmit(d0, sig3, sig3.duration);
+        transmit(d1, sig4, sig4.duration);
+    }
+
+    qubit $0;
+    q0_cal1 $0;
+    // Implicit barrier brings both clocks to ``100dt`` (``sig1.duration``) at start of ``q0_cal2``
+    q0_cal2 $0;
+    // Implicit barrier brings both clocks to ``175dt`` (``lengthof(q0_cal1)+sig4.duration``) at start of next ``defcal``
+
+``defcal`` blocks must have a well-defined length for all possible input argument combinations and must be provided
+to the scheduler at the circuit level.
+
+It is critical that the update time of carrier properties be well-defined such that the appropriate absolute phase may be
+accumulated. Updates of carrier properties are defined to occur at the maximum time across all channels used up to that point.
```

Comment:
> the maximum time across all channels used up to that point 

Is this actually knowable when carriers are shared across `defcals` and there is control flow?